### ISSUE TITLE
Consolidate fingerprint collection efforts with citizenlab/filtering-annotations

### ISSUE THREAD

**Issue by hellais** [ooni/backend#516]
I have a bunch of iPython notebooks that include a list of blockpages and reference an OONI measurements.

I have not done a great job at consolidating all this knowledge in one place, so I will use this ticket as such a place.
I am sorry that some don't include the source measurement as reference, but that should be fairly easy to mine from the data.

The `CENSOR_IP` array includes a list of IPs that are known to be used by censors when spoofing DNS responses.

* [x] Malaysia
```
CENSOR_IPS = [
    "175.139.142.25"
]
BLOCKPAGE_FINGERPRINTS = [
    "This website is not available in Malaysia as it violate"
]
```

* [x] Egypt

```
EG_BLOCKPAGES = [
    {
        'match': [
            (
                'body_s_re',
                 re.compile(b'^<HTML></HTML>$')
            )
        ],
        'block_id': 'eg-vodafone-0',
        'measurement_id': [
        ]
    }
]
EG_CENSOR_IPS = []
```

* [x] Indonesia
```
ID_BLOCKPAGES = [
        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'Apabila Anda merasa situs yang dikunjungi tidak bertentangan dengan program pemerintah, silakan hubungi customer customer support kami.')
                )
            ],
            'block_id': 'id-bold-4g-0',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/id-bolt-4g.html'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'<title>FastNet | Internet Sehat dan Aman</title>')
                )
            ],
            'block_id': 'id-first-0',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/id-first.html'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'negatif sesuai dengan Peraturan Menteri Kominfo No 19 Tahun 2014')
                )
            ],
            'block_id': 'id-indosat-0',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/id-indosat-2-Oct2014-update.htm'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'<b>Access Restricted by</b>')
                )
            ],
            'block_id': 'id-indosat-1',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/id-indosat.html'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'tidak bisa diakses melalui jaringan ini sesuai peraturan perundang-undangan')
                )
            ],
            'block_id': 'id-jalawave-0',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/id-jalawave.html'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'<title>Internet Positif - Indonesia</title>')
                )
            ],
            'block_id': 'id-telekomunikasi-0',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/id-telkomunikasi.html'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'<p>which is not permitted by goverment regulation.</p>')
                )
            ],
            'block_id': 'id-xl-0',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/id-xl-netsweeper.html'
            ]
        },

        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'<TITLE>ERROR BLOCKED</TITLE>')
                )
            ],
            'block_id': 'id-xl-1',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/id-xl.html'
            ]
        },
    
        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'replace\("http://internetpositif.uzone.id/page')
                )
            ],
            'block_id': 'id-positif-0',
            'measurement_id': [
                'https://explorer.ooni.torproject.org/measurement/20161218T000138Z_AS17974_MjcGCXBqEicdyD6UzGEaD8rkYuBTuPIeqiqChtLmLo2catzKsP?input=http:%2F%2Fwww.ashemaletube.com'
            ]
        },      
    ]
ID_CENSOR_IPS = [
    '36.86.63.185',
    '112.215.37.181', 
    '114.4.66.127',
    '118.98.96.151',
    '118.97.116.27',
    '180.214.232.61',
    '202.3.219.209',
    '202.62.29.1',
    '202.73.99.3', # https://explorer.ooni.torproject.org/measurement/20161226T000032Z_AS17974_fpJd2uwEJTW9uGSvFFQ64zrxuMsF6OEp24wTl6nIeeu5EugGX9?input=http:%2F%2Fwww.ashemaletube.com
    '210.210.178.70'
]
```

* [x] Thailand
```
TH_BLOCKPAGES = [
        {
            'match': [
                (
                    'body_s_re',
                     re.compile('<h1>เว็บไซต์นี้มีเนื้อหาและข้อมูลที่ไม่เหมาะสม<BR>ถูกระงับโดยกระทรวงดิจิทัลเพื่อเศรษฐกิจและสังคม</h1>'.encode('utf-8'))
                )
            ],
            'block_id': 'th-ministry-0',
            'measurement_id': [
                '20170214T023413Z_AS132061_uOj4NMFvbADZII8Sp2cmuZbgnozyuePra6Dyp5i16tQRblJYDW', 
                'http://www.wikileaks.org'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'<iframe src="http://103.208.24.21/')
                )
            ],
            'block_id': 'th-ministry-1',
            'measurement_id': [
                '20170210T012335Z_AS24378_8qwIdZPybKyYfAhU4RswGjL8CpdDjLs91F0YtCRo8IQf4gIJZp', 
                'http://xhamster.com'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile('กระทรวงเทคโนโลยีสารสนเทศและการสื่อสาร ได้มีคำสั่งให้ระงับการเผยแพร่เว็บไซด์ที่ท่านต้องการเข้าชม'.encode('utf-8'))
                )
            ],
            'block_id': 'th-ministry-2',
            'measurement_id': [
                '20170222T131301Z_AS45758_Np4oEy8rsv8s9C9g1Xs1r00zm1xqPIthWls9phCG3SdVuPuL1A', 
                'http://www.naughty.com'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile('เว็บไซต์ที่ท่านต้องการเข้าชมได้ถูกระงับการเผยแพร่ตามคำสั่งจากกระทรวงเทคโนโลยีสารสนเทศและการสื่อสาร'.encode('utf-8'))
                )
            ],
            'block_id': 'th-ministry-3',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/th-martial_law2014/tot-court-announce.html'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile('The Ministry of Information and Communication Technology has temporarily ceased the service to access such kind of information'.encode('utf-8'))
                )
            ],
            'block_id': 'th-ministry-4',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/th-martial_law2014/tot-ict.html'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile('<center> <img src="ict.jpg"> </center>'.encode('utf-8'))
                )
            ],
            'block_id': 'th-ministry-5',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/th-martial_law2014/tot-june4th-cop-ict.html'
            ]
        },       
        {
            'match': [
                (
                    'body_s_re',
                     re.compile('กองบังคับการปราบปรามการกระทําความผิดเกี่ยวกับอาชญากรรมทางเทคโนโลยี'.encode('utf-8'))
                )
            ],
            'block_id': 'th-cyber-0',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/th-martial_law2014/3bb-tcsd.html',
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile('An  access to such information has been temporarily ceased </span>'.encode('utf-8'))
                )
            ],
            'block_id': 'th-temp-0',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/th-martial_law2014/tot-temportary-cease.html',
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile('<center> <img src="./Redirecting_files/ict.jpg"> </center>'.encode('utf-8'))
                )
            ],
            'block_id': 'th-ministry-6',
            'measurement_id': [
                'https://github.com/NullHypothesis/blockpages/blob/master/TH/216.36.238.43/Redirecting.html',
            ]
        },        
        {
            'match': [
                (
                    'body_s_re',
                     re.compile('ตามคำสั่ง  กองอำนวยการรักษาความสงบเรียบร้อย'.encode('utf-8'))
                )
            ],
            'block_id': 'th-peace-corp-0',
            'measurement_id': [
                'https://github.com/citizenlab/blockpages/blob/master/html/th-martial_law2014/3bb-maintain_order.html'
            ]
        },
        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'<meta HTTP-EQUIV="REFRESH" content="0; url=http://192.168.1.1/blocking')
                )
            ],
            'block_id': 'th-local-block-0',
            'measurement_id': [
                '20161218T000023Z_AS45758_owGKv7as3ueVKmfKpNfRnsDJGOsahdKgwDw0X301iaTBMlqHRt', 
                'http://crackspider.net'
            ]
        }
    ]
```

* [x] Uganda
```
UG_BLOCKPAGES = [
        {
            'match': [
                (
                    'body_s_re',
                     re.compile(b'Cyberoam is securing you. Click here to know more.')
                )
            ],
            'block_id': 'ug-cyber-roam-0',
            'measurement_id': [
                '20160927T035653Z_AS37027_3rGH0c3HxdARB8yPZbQIY9GevVO5FPLESGqrAojrg6XqYemfBZ', 
                'http://redtube.com'
            ]
        },
    ]
```

As I find more ipython notebooks I will edit the ticket. Please add a comment to this ticket you have imported some of these.

**Comment by hellais** [comment]
So I started going through this and then realised that I think we are tackling this in the wrong way.

It seems like there is some duplication of effort with the work that the @jakubd has been doing with the https://github.com/citizenlab/filtering-annotations repo.

He's been doing an excellent job at documenting and keeping the fingerprints up to date and I recon that we should rather join forces with them so we can have a single source of truth regarding blockpage fingerprints.

I went through some of the fingerprints above and most of them were already present in filtering-annotations, yet a few were missing.

I also did some cherry pick comparisons between the fingerprints we have in https://github.com/ooni/pipeline/blob/master/af/fastpath/fastpath/utils.py#L17 and there are for sure some fingerprints which we have that are not part of filtering-annotations and vice-versa.

My suggestion is that we use the filtering-annotations repo as the authoritative source for blockpages and DNS fingerprints and upstream the fingerprints that are missing from here.

The challenges that I see in doing this are the following:
1. It's not so simple to do a diff between the fingerprints we have vs those that they have. This is because the same fingerprint may be defined using a different pattern matching rule and so we run the risk of duplicating them.
2. There are some small differences in the logic we use for defining the fingerprint patterns. For example filtering-annotations does a full text search inside of all headers, while we only search for the prefix in a specified header name.

My ideas for addressing these two problems are as follows:
1. We add support to the pipeline for matching against both the fingerprints inside of the fastpath and additionally those in the citizenlab repo. The matched fingerprints should be written to a specific column and we should enumerate all the matching fingerprints. At the end of the run we will have some that matched the citizenlab fingerprint db, some of them only ours and some which are duplicate matching both. We will then take only the ones that are missing from the citizenlab repo (i.e. that matched only our fingerprint db) and upstream them. Once this is done, we will change our code to use the citizenlab repo as the canonical source.
2. I think we should change the logic of our fingerprint matching rules to be the same as the one used by the citizenlab repo. I also think we should record, when we hit a fingerprint match, the ID of the fingerprint that was matched against as it's quite useful metadata.

I am going to rename this issue to reflect the changed scope.
